internal/cmdrender/runtime/container.go

```diff
@@ -58,27 +89,38 @@ type ContainerFn struct {
 // It reads the input from the given reader and writes the output
 // to the provided writer.
 func (f *ContainerFn) Run(reader io.Reader, writer io.Writer) error {
-	pr := printer.FromContextOrDie(f.Ctx)
+	// check and pull image before running to avoid polluting CLI
+	// output
+	err := f.prepareImage()
+	if err != nil {
+		return fmt.Errorf("failed to check function existence: %w", err)
+	}
+
 	errSink := bytes.Buffer{}
 	cmd, cancel := f.getDockerCmd()
 	defer cancel()
 	cmd.Stdin = reader
 	cmd.Stdout = writer
 	cmd.Stderr = &errSink
 
-	pr.PkgPrintf(f.Path, "running function %q: ", f.Image)
 	if err := cmd.Run(); err != nil {
-		pr.Printf("FAILED\n")
-		return fmt.Errorf("%s", errSink.String())
+		var exitCode int
+		if exitError, ok := err.(*exec.ExitError); ok {
```

Comment:
tip: use `errors.As(..)` to make it idiomatic.

Reply:
done